packages/protocol/contracts/governance/VestingInstance.sol

```diff
@@ -0,0 +1,332 @@
+    uint256 withdrawableAmount = getWithdrawableAmountAtTimestamp(revokeTime);
+    uint256 refundAmount = balance.sub(withdrawableAmount);
+    revoked = true;
+    require(
+      getGoldToken().transfer(refundDestination, refundAmount),
+      "Transfer of refund upon revokation failed"
+    );
+    emit VestingRevoked(revokeTime);
+  }
+
+  /**
+   * @notice Allows only the revoker to pause the gold withdrawal
+   * @param pausePeriod the period for which the withdrawal shall be paused
+   */
+  function pause(uint256 pausePeriod) external onlyRevoker {
+    require(!paused, "Vesting withdrawals already paused");
+    require(revocable, "Vesting must be revokable");
+    require(!revoked, "Vesting already revoked");
+    require(pausePeriod <= 365 days, "Pause period is limited to max. 365 days");
+    paused = true;
+    pauseEndTime = block.timestamp.add(pausePeriod);
+    emit WithdrawalPaused(pausePeriod, block.timestamp);
+  }
+
+  /**
+   * @dev Calculates the amount that has already vested but hasn't been withdrawn yet.
+   * @param timestamp the timestamp at which to calculate the withdrawable amount
+   * @return The withdrawable amount at the timestamp
+   * @dev Function is also made public in order to be called for informational purpose
+   */
+  function getWithdrawableAmountAtTimestamp(uint256 timestamp) public view returns (uint256) {
+    return calculateVestedAmountAtTimestamp(timestamp).sub(currentlyWithdrawn);
+  }
+
+  /**
+   * @notice Calculates both non-locked and locked gold balance parts and returns their sum as the total balance of the vesting instance
+   * @return The total vesting instance balance
+   */
+  function getVestingInstanceTotalBalance() public view returns (uint256) {
+    uint256 nonLockedBalance = getGoldToken().balanceOf(address(this));
+    uint256 lockedBalance = getLockedGold().getAccountTotalLockedGold(address(this));
+    return nonLockedBalance.add(lockedBalance);
+  }
+
+  /**
+   * @dev Calculates the amount that has already vested.
+   * @param timestamp the timestamp at which to calculate the already vested amount
+   */
+  function calculateVestedAmountAtTimestamp(uint256 timestamp) private view returns (uint256) {
+    uint256 currentBalance = getVestingInstanceTotalBalance();
+    uint256 totalBalance = currentBalance.add(currentlyWithdrawn);
+    uint256 vestingPeriods = vestingSchedule.vestingAmount.div(vestingSchedule.vestAmountPerPeriod);
+
+    if (timestamp < vestingSchedule.vestingCliffStartTime) {
+      return 0;
+    }
+    if (
+      timestamp >=
+      vestingSchedule.vestingStartTime.add(vestingPeriods.mul(vestingSchedule.vestingPeriodSec)) ||
+      revoked
+    ) {
+      return totalBalance;
+    }
+
+    uint256 timeSinceStart = timestamp.sub(vestingSchedule.vestingStartTime);
+    uint256 timeToEnd = vestingPeriods.mul(vestingSchedule.vestingPeriodSec);
+    return totalBalance.mul(timeSinceStart).div(timeToEnd);
```

Comment:
This is not correct. We are treating vesting as a continuoum, when it's discrete. (eg. i vest on discrete amount of time, not on every second).

So computing a % of vesting should be done in "Floor(amounts of periods)" not in "amount of seconds"